### PR TITLE
Fixed clean up old compact chunks when compact starts

### PR DESCRIPTION
**Pull request by jojohappy**
Signed-off-by: jojohappy <sarahdj0917@gmail.com>

## Changes

fixes: #470 

Related discussion: [comment](https://github.com/improbable-eng/thanos/issues/470#issuecomment-426691885)

Clean up the temporary directory before compactor beginning.

### CODE REVIEW

**Comment by domgreen** [comment]
We can actually do it as soon as we go into the compact method as we do not care at all about the existing blocks when we run a compaction loop.

**Comment by domgreen** [comment]
I would also add clearing up of the downsample dir in this pr somewhere as this would have pulled down all the compacted blocks. It will clean it up when it starts a downsample run but if the machine is already done that it may not succeed in doing the original compaction.

**Comment by jojohappy** [comment]
@domgreen Do you mean we should clean up the temporary directory before every compaction loop?

I have check the code that the downsample temporary dir will be cleaned at the beginning of processing downsampling.

**Comment by jojohappy** [comment]
@domgreen Thank your for your detail explain. I have changed my code to clean up the temporary directory at the beginning of every compaction loop.

**Comment by jojohappy** [comment]
@domgreen I have rollbacked the code and clean up the downsampling temporary directory before processing compaction. Does it make sense?

File: cmd/thanos/compact.go [hunk 1]
```diff
@@ -147,6 +148,9 @@ func runCompact(
 
 	ctx, cancel := context.WithCancel(context.Background())
 	f := func() error {
+		if err := os.RemoveAll(downsamplingDir); err != nil {
```
Comment (domgreen) [inline cmd/thanos/compact.go]: lets move this up to line `136` and change it to `os.RemoveAll(dataDir)` and we are good to go

**Comment by jojohappy** [comment]
Yes, I have tested that it will remove the temporary dir itself. (:

**Comment by xjewer** [comment]
it breaks backward compatibility:

```
compact command failed: clean working temporary directory: remove /var/thanos/data: device or resource busy
```

I use `data-dir` the same as the mount path of volume in k8s, hence dir couldn't be deleted.
The workaround is to set `data-dir` as nested dir. I.e. in my case this's: `/var/thanos/data/compact`

**Comment by domgreen** [comment]
Urgh, missed that case :( let me add a ticket and we can get someone to fix asap.

**Pull request merged**